pkg/ccl/backupccl/backup_test.go

```diff
@@ -10303,3 +10305,293 @@ func TestBackupDBWithViewOnAdjacentDBRange(t *testing.T) {
+		return nil
+	})
+	require.NoError(t, err)
+
+	// Bakcup the whole src cluster.
+	srcSQLDB.Exec(t, `BACKUP INTO $1`, localFoo)
+
+	// Restore into target cluster.
+	tgtSQLDB.Exec(t, `RESTORE FROM LATEST IN $1`, localFoo)
+	tgtSQLDB.Exec(t, `USE db1`)
+
+	// Verify that all IDs are correctly rewritten.
+	rows = tgtSQLDB.QueryStr(t, `SELECT function_id FROM crdb_internal.create_function_statements WHERE function_name = 'f1'`)
+	require.Equal(t, 1, len(rows))
+	require.Equal(t, 1, len(rows[0]))
+	udfID, err = strconv.Atoi(rows[0][0])
+	require.NoError(t, err)
+	err = sql.TestingDescsTxn(ctx, tgtServer, func(ctx context.Context, txn *kv.Txn, col *descs.Collection) error {
+		dbDesc, err := col.GetImmutableDatabaseByName(ctx, txn, "db1", tree.DatabaseLookupFlags{Required: true})
+		require.NoError(t, err)
+		require.Equal(t, 107, int(dbDesc.GetID()))
+
+		scDesc, err := col.GetImmutableSchemaByName(ctx, txn, dbDesc, "sc1", tree.SchemaLookupFlags{Required: true})
+		require.NoError(t, err)
+		require.Equal(t, 125, int(scDesc.GetID()))
+
+		tbName := tree.MakeTableNameWithSchema("db1", "sc1", "tbl1")
+		_, tbDesc, err := col.GetImmutableTableByName(
+			ctx, txn, &tbName, tree.ObjectLookupFlagsWithRequired(),
+		)
+		require.NoError(t, err)
+		require.Equal(t, 126, int(tbDesc.GetID()))
+
+		typName := tree.MakeQualifiedTypeName("db1", "sc1", "enum1")
+		_, typDesc, err := col.GetImmutableTypeByName(ctx, txn, &typName, tree.ObjectLookupFlagsWithRequired())
+		require.NoError(t, err)
+		require.Equal(t, 127, int(typDesc.GetID()))
+
+		tbName = tree.MakeTableNameWithSchema("db1", "sc1", "sq1")
+		_, tbDesc, err = col.GetImmutableTableByName(
+			ctx, txn, &tbName, tree.ObjectLookupFlagsWithRequired(),
+		)
+		require.NoError(t, err)
+		require.Equal(t, 129, int(tbDesc.GetID()))
+
+		fnDesc, err := col.GetImmutableFunctionByID(ctx, txn, descpb.ID(udfID), tree.ObjectLookupFlagsWithRequired())
+		require.NoError(t, err)
+		require.Equal(t, 130, int(fnDesc.GetID()))
+		require.Equal(t, 107, int(fnDesc.GetParentID()))
+		require.Equal(t, 125, int(fnDesc.GetParentSchemaID()))
+		// Make sure db name and IDs are rewritten in function body.
+		require.Equal(t, "SELECT a FROM db1.sc1.tbl1;\nSELECT nextval(129:::REGCLASS);", fnDesc.GetFunctionBody())
+		require.Equal(t, 100127, int(fnDesc.GetArgs()[0].Type.Oid()))
+		require.Equal(t, []descpb.ID{126, 129}, fnDesc.GetDependsOn())
+		require.Equal(t, []descpb.ID{127, 128}, fnDesc.GetDependsOnTypes())
+
+		fnDef, _ := scDesc.GetFunction("f1")
+		require.Equal(t, 130, int(fnDef.Overloads[0].ID))
+		require.Equal(t, 100127, int(fnDef.Overloads[0].ArgTypes[0].Oid()))
+		return nil
+	})
+	require.NoError(t, err)
+
+	// Make sure function actually works on the target cluster.
+	rows = tgtSQLDB.QueryStr(t, `SELECT sc1.f1('Good'::sc1.enum1)`)
+	require.Equal(t, 1, len(rows))
+	require.Equal(t, 1, len(rows[0]))
+	require.Equal(t, "1", rows[0][0])
+	require.NoError(t, err)
+
+}
```

Comment:
Is there any reason why these tests can't be implemented as data-driven tests? See `TestDataDriven`. This would be much more preferable. Also, I wouldn't count too much on individual ID values being stable over time, but that's a secondary concern.

Reply:
I was a fool and didn't notice that's a thing for backup/restore :( Will move these tests there and add more tests.

Reply:
Ignorance is not foolishness.